src/app/clusters/time-synchronization-server/time-synchronization-server.cpp

```diff
@@ -0,0 +1,685 @@
+        return false;
+    }
+
+    // re-schedule event for after min 1hr if no time is still available
+    ChipLogProgress(Zcl, "Emit sendTimeFailure event [endpointId=%d]", endpointId);
+    return true;
+}
+
+static bool sendMissingTrustedTimeSource(chip::EndpointId endpointId)
+{
+    Events::MissingTrustedTimeSource::Type event;
+    EventNumber eventNumber;
+
+    CHIP_ERROR error = LogEvent(event, endpointId, eventNumber);
+
+    if (CHIP_NO_ERROR != error)
+    {
+        ChipLogError(Zcl, "Unable to send sendMissingTrustedTimeSource event [endpointId=%d]", endpointId);
+        return false;
+    }
+
+    // re-schedule event for after min 1hr if TTS is null or cannot be reached
+    ChipLogProgress(Zcl, "Emit sendMissingTrustedTimeSource event [endpointId=%d]", endpointId);
+    return true;
+}
+
+static void utcTimeChanged(uint64_t utcTime)
+{
+    System::Clock::Seconds32 lastKnownGoodChipEpoch;
+    System::Clock::Microseconds64 realTime;
+    uint32_t utcTimetoChipEpoch;
+
+    Server::GetInstance().GetFabricTable().GetLastKnownGoodChipEpochTime(lastKnownGoodChipEpoch);
+    System::SystemClock().GetClock_RealTime(realTime);
+    ChipLogError(Zcl, "UTCTime: %llu Last Known Good Time: %u Real Time: %llu", utcTime, lastKnownGoodChipEpoch.count(),
+                 realTime.count());
+    chip::UnixEpochToChipEpochTime((uint32_t)(utcTime / chip::kMicrosecondsPerSecond), utcTimetoChipEpoch);
+
+    if (utcTimetoChipEpoch >= lastKnownGoodChipEpoch.count()) // update Last Known Good Time if a more recent time is obtained
+    {
+        Server::GetInstance().GetFabricTable().SetLastKnownGoodChipEpochTime(System::Clock::Seconds32(utcTimetoChipEpoch));
+        System::SystemClock().SetClock_RealTime(System::Clock::Microseconds64(utcTime));
+
+        Server::GetInstance().GetFabricTable().GetLastKnownGoodChipEpochTime(lastKnownGoodChipEpoch);
+        System::SystemClock().GetClock_RealTime(realTime);
+        ChipLogError(Zcl, " Last Known Good Time: %u Real Time: %llu", lastKnownGoodChipEpoch.count(), realTime.count());
+    }
+}
+
+bool emberAfTimeSynchronizationClusterSetUtcTimeCallback(
+    chip::app::CommandHandler * commandObj, const chip::app::ConcreteCommandPath & commandPath,
+    const chip::app::Clusters::TimeSynchronization::Commands::SetUtcTime::DecodableType & commandData)
+{
+    Optional<StatusCode> status = Optional<StatusCode>::Missing();
+    Status globalStatus         = Status::Success;
+
+    auto utcTime     = commandData.utcTime;
+    auto granularity = commandData.granularity;
+    auto timeSource  = commandData.timeSource;
+
+    TimeSynchronization::GranularityEnum currentGranularity;
+    Granularity::Get(commandPath.mEndpointId, &currentGranularity);
+
+    if (currentGranularity == TimeSynchronization::GranularityEnum::kNoTimeGranularity || granularity >= currentGranularity)
+    {
+        UTCTime::Set(commandPath.mEndpointId, utcTime);
```

Comment:
This is going to trigger subscription updates, which is a spec violation.

I would have expected us to not store this in the attribute store at all.

Reply:
quality C

Reply:
Yes, spec has quality C.  And the way that's implemented in the SDK is _you don't put those things in the attribute store and never mark them dirty_.

As in, there is no actual support for C in the SDK at the moment, other than "don't notify anyone if the value changes".